distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso

```diff
@@ -656,6 +658,113 @@ type Column
+
+       ? Negative decimal place counts
+         Rounding to `n` digits can be thought of as "rounding to the nearest
+         multiple of 10^(-n)". For negative decimal counts, this results in
+         rounding to the nearest positive integer power of 10.
+
+       > Example
+         Round a column of `Decimal` values`.
+
+             Column.from_vector "foo" [1.2, 2.3, 3.6] . round == (Column.from_vector "foo" [1, 2, 4])
+    round : Integer -> Boolean -> Problem_Behavior -> Column | Illegal_Argument | Invalid_Value_Type
+    round self decimal_places=0 use_bankers=False on_problems=Report_Warning = Value_Type.expect_numeric self <|
+        # If it's an integer column and decimal_places >=0 then it's a no-op.
+        if self.value_type.is_integer && decimal_places >= 0 then self else
+            returns_double = decimal_places > 0
+            builder = if returns_double then make_double_builder else make_long_builder
+            fun = _.round decimal_places use_bankers
+            Column_Ops.map_over_storage self fun builder skip_nothing=True on_problems=on_problems
+
+    ## ALIAS int
+
+       If the column is numeric, truncate the floating-point values to an
+       integer by dropping the fractional part. This is equivalent to
+       "round-toward-zero". If the column is of type `Date_Time`, truncates the
+       values to `Date`.
+
+       > Example
+         Truncate a column of `Decimal` values.
+
+             Column.from_vector "foo" [1.25, 2.33, 3.57] . truncate == (Column.from_vector "foo" [1, 2, 3])
+
+       > Example
+         Truncate a column of `Date_Time` values.
+            date_times = Column.from_vector "foo" [Date_Time.new 2020 10 24 1 2 3, Date_Time.new 2020 10 24 1 2 3]
+            dates = Column.from_vector "foo" [Date.new 2020 10 24, Date.new 2020 10 24]
+            col.truncate == dates
+    truncate : Column ! Invalid_Value_Type
+    truncate self =
+        case self.value_type.is_numeric of
+            True ->
+                fun = _.truncate
+                Column_Ops.map_over_storage self fun make_long_builder skip_nothing=True
+            False -> case self.value_type == Value_Type.Date_Time of
+                True ->
+                    fun = _.date
+                    Column_Ops.map_over_storage self fun make_date_builder_adapter skip_nothing=True
+                False -> Invalid_Value_Type.Column "Numeric or Date_Time" self.value_type
+
+    ## Computes the nearest integer above this number for values in a numeric
+       column.
+
+       Returns a column of `Integer`.
+
+       > Example
+         Take the ceiling of a column of `Decimal` values.
+
+             Column.from_vector "foo" [1.25, 2.33, 3.57] . ceil == (Column.from_vector "foo" [2, 3, 4])
+    ceil : Column ! Invalid_Value_Type
+    ceil self = Value_Type.expect_numeric self <|
+        fun = _.ceil
+        Column_Ops.map_over_storage self fun make_long_builder skip_nothing=True
```

Comment:
I'm really not sure if these operations should be implemented in Enso.

Currently most of the 'primitive' operations, especially on numeric columns are implemented in Java, for better efficiency. 

Do we want to move towards doing these in Enso? I think we should measure what is the difference in performance.

Reply:
Another issue.

Currently I see that the `ceil` method may return an EnsoBigInteger (refer to the `decimal/CeilNode.java`).

However, currently the Table library storage _does not support_ Big Integer storage - only `LongStorage`. So this will likely fail in some way.

I think we need to have tests checking what happens in the scenario where the `double` value magnitude exceeds the `long` range, to verify the behaviour.

IMO currently it could be valid to return `Nothing` in such a case and attach some kind of `Arithmetic_Overflow` exception. Extending Table to handle Big Integers is a separate issue (currently unscheduled, @jdunkerley do we have plans to support this?)

Reply:
To be clear about the first issue. The `map_over_storage` is OK. Its usage is warranted when the operation we perform on each row must run some Enso logic (some complicated Enso function that is not worth replicating into Java, or has so expensive logic anyway the benefit of inlining it to Java would be negligible).

But I think we shouldn't use Enso operations for primitive operations and instead implement them directly in Java. That is the current architecture of the primitive-value storages in the Table library. We can discuss if it should be changed, but to do so I think we need to do some performance measurements to take informed decisions.

Reply:
I think worth making this a vectorised operation on the NumberStorage.

We can use the Core_Utils to share the common implementation. Lets put it as a follow up ticket though please.

Reply:
> We can use the Core_Utils to share the common implementation. Lets put it as a follow up ticket though please.

We could make the `round` implementation in `Core_Utils`, but then it would no longer be able to be written in Enso...

As for all the others - `ceil`, `floor` and I think `truncate` - we seem to be doing just the basic Java operations `Math.ceil` etc. in the 'primitive' path. The only additional handling is to handle promotion to `BigInteger` if the result exceeds the range of `long`. But `BigInteger` is currently not supported by Table anyway, so I don't think sharing the implementation of these gives us any value _at this moment_ - because the pure Enso side needs the additional BigInteger handling that has some Truffle magic, and the Table side needs to just report a warning/failure in case of an overflow - the common part is _just_ delegating to `Math.ceil` and co. which IMO isn't enough to make sharing worth it.

(But I'd make these vectorized nonetheless, I'm only unsure about `round` as that would require us moving the whole implementation into Java - it could make it more efficient for tables though, so may as well be worth it, but it's quite a bit more work).

Reply:
Ok - not need to put in `Core_Utils` then.

Worth us seeing the relative performance of rounding 1,000,000 values as a map vs a vectorised op.